src/Utility/ToolsUtility.cs

```diff
@@ -0,0 +1,194 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using System.Text.Json;
+using System.Threading.Tasks;
+using OpenAI.Agents;
+using OpenAI.Embeddings;
+
+namespace OpenAI;
+
+internal static class ToolsUtility
+{
+    internal static readonly BinaryData EmptyParameters = BinaryData.FromString("""{ "type" : "object", "properties" : {} }""");
+    internal const string McpToolSeparator = "_-_";
+
+    internal static string GetMethodDescription(MethodInfo method)
+    {
+        var description = method.Name;
+        var attr = method.GetCustomAttribute<DescriptionAttribute>();
+        if (attr != null)
+            description = attr.Description;
+        return description;
+    }
+
+    internal static string GetParameterDescription(ParameterInfo param)
+    {
+        string description = param.Name!;
+        var attr = param.GetCustomAttribute<DescriptionAttribute>();
+        if (attr != null)
+            description = attr.Description;
+        return description;
+    }
+
+    internal static ReadOnlySpan<byte> ClrToJsonTypeUtf8(Type clrType) =>
+        clrType switch
+        {
+            Type t when t == typeof(double) => "number"u8,
+            Type t when t == typeof(int) => "number"u8,
+            Type t when t == typeof(string) => "string"u8,
+            Type t when t == typeof(bool) => "bool"u8,
+            _ => throw new NotImplementedException()
+        };
+
+    internal static string ClrToJsonTypeUtf16(Type clrType) =>
+        clrType switch
+        {
+            Type t when t == typeof(double) => "number",
+            Type t when t == typeof(int) => "number",
+            Type t when t == typeof(string) => "string",
+            Type t when t == typeof(bool) => "bool",
+            _ => throw new NotImplementedException()
+        };
+
+    internal static BinaryData BuildParametersJson(ParameterInfo[] parameters)
+    {
+        if (parameters.Length == 0)
+            return EmptyParameters;
+
+        var required = new List<string>();
+        using var stream = new MemoryStream();
+        using var writer = new Utf8JsonWriter(stream);
+        writer.WriteStartObject();
+        writer.WriteString("type"u8, "object"u8);
+        writer.WriteStartObject("properties"u8);
+
+        foreach (ParameterInfo parameter in parameters)
+        {
+            writer.WriteStartObject(parameter.Name!);
+            writer.WriteString("type"u8, ClrToJsonTypeUtf8(parameter.ParameterType));
+            writer.WriteString("description"u8, GetParameterDescription(parameter));
+            writer.WriteEndObject();
+
+            if (!parameter.IsOptional || (parameter.HasDefaultValue && parameter.DefaultValue is not null))

```

Comment:
Curious: Why is an optional parameter with a non-null default considered required?

Reply:
Because it effectively is always included with at least a default value.